### PR TITLE
Arm stencils

### PR DESCRIPTION
**Pull request by noraabiakar**
This PR contains the files I have added/changes I have made in the context of the arm study for gridtools. 

- Added 3 new stencil variations (.cpp/.h)
src/knl/knl_hdiff_variant_ij_private_blocks
src/knl/knl_vadv_variant_ik_blocked_j
src/knl/knl_vadv_variant_ij_blocked_i_vector

-I did not create a new subdirectory for thunderX2 stencils. All stencil files are under src/knl. All compile and run on arm. 

-Added a sbench_tx2 target to the makefile, it compiles the target with the correct cache size for the cache flusher and the right page size for alignment. 

-Added small functionalities and fixes.

### CODE REVIEW

File: src/hdiff_stencil_variant.h [hunk 1]
```diff
@@ -20,6 +20,7 @@ namespace platform {
 
         std::vector<std::string> stencil_list() const override;
 
+        void prerun_init() override;
```
Comment (fthaler) [inline src/hdiff_stencil_variant.h]: Why do we need this? Can’t this work be done in the constructor?

File: src/knl/knl_hdiff_variant_ij_private_blocks.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+#pragma omp parallel 
+			{
+				int thread_id = omp_get_thread_num();
+				value_type *__restrict__ lap = lap_data.data() + thread_id * (m_iblocksize + 2) * (m_jblocksize + 2);
+				value_type *__restrict__ flx = flx_data.data() + thread_id * (m_iblocksize + 1) * (m_jblocksize);
+				value_type *__restrict__ fly = fly_data.data() + thread_id * (m_iblocksize) * (m_jblocksize + 1);
+#pragma omp for collapse(3)
+				for (int jb = 0; jb < jsize; jb += m_jblocksize) {
+			        for (int k = 0; k < ksize; ++k) {
```
Comment (fthaler) [inline src/knl/knl_hdiff_variant_ij_private_blocks.cpp]: Indentation is off in many places, please format using clang-format if possible (there’s a `make format` target actually).

File: src/knl/knl_vadv_variant_ij_blocked.cpp [hunk 3]
```diff
@@ -25,11 +25,9 @@ namespace platform {
             const int isize = this->isize();
             const int jsize = this->jsize();
             const int ksize = this->ksize();
-            constexpr int istride = 1;
+            const int istride = this->istride();
```
Comment (fthaler) [inline src/knl/knl_vadv_variant_ij_blocked.cpp]: Not sure if we can do that without losing ICC’s vectorization, will check…

File: src/knl/knl_vadv_variant_ij_blocked_i_vector.h [hunk 4]
```diff
@@ -0,0 +1,188 @@
+#pragma once
+
+#include "knl/knl_platform.h"
+#include "knl/knl_vadv_variant.h"
+#include "omp.h"
+
+#define eproma 64
```
Comment (fthaler) [inline src/knl/knl_vadv_variant_ij_blocked_i_vector.h]: Use constexpr constant instead of macro?

File: src/knl/knl_vadv_variant_ik_blocked_j.h [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+            vadv_variant_ik_blocked_j(const arguments_map &args)
+                : knl_vadv_stencil_variant<ValueType>(args), m_iblocksize(args.get<int>("i-blocksize")),
+                  m_jblocksize(args.get<int>("j-blocksize")) {
+                if (m_iblocksize <= 0 || m_jblocksize <= 0)
+                    throw ERROR("invalid block size");
+				//set sizes of caches 
+				int num_threads; 
+#pragma omp parallel 
+				{
+					num_threads = omp_get_num_threads();
```
Comment (fthaler) [inline src/knl/knl_vadv_variant_ik_blocked_j.h]: Just use omp_get_max_threads().

File: src/knl/knl_vadv_variant_ik_blocked_j.h [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+                const int istride,
+                const int jstride,
+                const int kstride, 
+				const int ib, 
+				const int jb) {
+
+                const int k = 0;
+                const int index = i * istride + j * jstride + k * kstride;
+                const int cache_index = (i - ib) + (j - jb) * m_iblocksize;
+                //const int cache_index = i * istride + j * jstride;
```
Comment (fthaler) [inline src/knl/knl_vadv_variant_ik_blocked_j.h]: Delete.

File: src/knl/knl_vadv_variant_ik_blocked_j.h [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+                const int ksize,
+                const int istride,
+                const int jstride,
+                const int kstride, 
+				const int ib, 
+				const int jb) {
+
+                const int index = i * istride + j * jstride + k * kstride;
+                const int cache_index = (i - ib) + (j - jb) * m_iblocksize;
+                //const int cache_index = i * istride + j * jstride;
```
Comment (fthaler) [inline src/knl/knl_vadv_variant_ik_blocked_j.h]: Delete.

File: src/knl/knl_vadv_variant_ik_blocked_j.h [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+                const int istride,
+                const int jstride,
+                const int kstride, 
+				const int ib, 
+				const int jb) {
+
+                const int k = ksize - 1;
+                const int index = i * istride + j * jstride + k * kstride;
+                const int cache_index = (i - ib) + (j - jb) * m_iblocksize;
+                //const int cache_index = i * istride + j * jstride;
```
Comment (fthaler) [inline src/knl/knl_vadv_variant_ik_blocked_j.h]: Delete.

File: src/result.cpp [hunk 9]
```diff
@@ -20,6 +21,14 @@ double result_array::max() const {
 
 double result_array::avg() const { return std::accumulate(m_data.begin(), m_data.end(), 0.0) / m_data.size(); }
 
+double result_array::stdev() const { 
+    std::vector<double> c_data = m_data;
```
Comment (fthaler) [inline src/result.cpp]: I think a for loop would be easier to follow here and eliminate the need of copying m_data:
```
double mean = avg();
double sum = 0;
for (double d : m_data)
    sum += (d - mean) * (d - mean);
return std::sqrt(sum / m_data.size());
```

File: Makefile [hunk 10]
```diff
@@ -1,53 +1,58 @@
...
 
 CXXVERSION=$(shell $(CXX) --version)
 ifneq (,$(findstring icpc,$(CXXVERSION)))
 	CXXFLAGS_KNL+=-xmic-avx512 -ffreestanding
 	CXXFLAGS_KNLCPU+=-ffreestanding
 else ifneq (,$(findstring g++,$(CXXVERSION)))
-	LIBS+=-lgomp
-	CXXFLAGS+=-Wno-unknown-pragmas -Wno-unused-variable
+	LIBS+=-lgomp 
+	CXXFLAGS+=-Wno-unknown-pragmas -Wno-unused-variable
```
Comment (fthaler) [inline Makefile]: Spaces added at the end of the line…

File: src/hdiff_stencil_variant.h [hunk 11]
```diff
@@ -83,39 +81,48 @@ namespace platform {
     }
 
     template <class Platform, class ValueType>
-    void hdiff_stencil_variant<Platform, ValueType>::prerun() {
+    void hdiff_stencil_variant<Platform, ValueType>::prerun_init() {
```
Comment (fthaler) [inline src/hdiff_stencil_variant.h]: Can’t we do this in the constructor?

File: src/hdiff_stencil_variant.h (hunk 1 above)
Comment (noraabiakar) [reply src/hdiff_stencil_variant.h]: We don't. It was easier to test the first-touch NUMA stuff this way. I moved it to the constructor.

File: src/knl/knl_hdiff_variant_ij_private_blocks.cpp (hunk 2 above)
Comment (noraabiakar) [reply src/knl/knl_hdiff_variant_ij_private_blocks.cpp]: Thanks, I hadn't noticed. I'll use clang-format

File: src/knl/knl_vadv_variant_ik_blocked_j.h (hunk 5 above)
Comment (noraabiakar) [reply src/knl/knl_vadv_variant_ik_blocked_j.h]: Thank you

File: src/knl/knl_vadv_variant_ik_blocked_j.h (hunk 6 above)
Comment (noraabiakar) [reply src/knl/knl_vadv_variant_ik_blocked_j.h]: Done.

File: src/knl/knl_vadv_variant_ik_blocked_j.h (hunk 7 above)
Comment (noraabiakar) [reply src/knl/knl_vadv_variant_ik_blocked_j.h]: Done.

File: src/knl/knl_vadv_variant_ik_blocked_j.h (hunk 8 above)
Comment (noraabiakar) [reply src/knl/knl_vadv_variant_ik_blocked_j.h]: Done.

File: src/result.cpp (hunk 9 above)
Comment (noraabiakar) [reply src/result.cpp]: I wanted to keep the same style as the previous functions. I agree though, this looks way better. I changed it.

File: Makefile (hunk 10 above)
Comment (noraabiakar) [reply Makefile]: Fixed.

File: src/hdiff_stencil_variant.h (hunk 11 above)
Comment (noraabiakar) [reply src/hdiff_stencil_variant.h]: Yep. Done.

File: src/knl/knl_vadv_variant_ij_blocked_i_vector.h (hunk 4 above)
Comment (noraabiakar) [reply src/knl/knl_vadv_variant_ij_blocked_i_vector.h]: I fixed it. Not entirely sure this is the correct way.

File: src/knl/knl_vadv_variant_ij_blocked.cpp (hunk 3 above)
Comment (fthaler) [reply src/knl/knl_vadv_variant_ij_blocked.cpp]: It actually kills the KNL performance…

Comment (noraabiakar) [reply src/knl/knl_vadv_variant_ij_blocked.cpp]: kij layout in this function behaves better than other layouts on ARM (we cant vectorize in any layout here). But still, this was not where we got the best vadv performance.

Comment (fthaler) [reply src/knl/knl_vadv_variant_ij_blocked.cpp]: It is (was) the fastest for KNL, though. So let’s just keep the constexpr thingy?

Comment (noraabiakar) [reply src/knl/knl_vadv_variant_ij_blocked.cpp]: Okay, done

**Pull request merged**